lib/spack/spack/solver/asp.py

```diff
@@ -1920,6 +1934,150 @@ def node_os(self, pkg, os):
     def node_target(self, pkg, target):
         self._arch(pkg).target = target
 
+    def error_conflict_triggered(self, msg):
+        raise UnsatisfiableSpecError(msg)
+
+    def error_no_version(self, pkg):
+        raise UnsatisfiableSpecError("No versions available for package '%s'" % pkg)
+
+    def error_versions_conflict(self, pkg, version1, version2):
+        msg = ("No version for '%s' satisfies '@%s' and '@%s'" %
+               (pkg, version1, version2))
+        raise UnsatisfiableSpecError(msg)
+
+    def error_version_unsatisfiable(self, pkg, constraint):
+        msg = "No valid version for '%s' satisfies '@%s'" % (pkg, constraint)
+        raise UnsatisfiableSpecError(msg)
+
+    def error_no_variant_value(self, pkg, variant):
+        msg = "No valid variant '%s' of package '%s'" % (variant, pkg)
+        raise UnsatisfiableSpecError(msg)
+
+    def error_multiple_values_sv_variant(self, pkg, variant, value1, value2):
+        variant1 = spack.spec.Spec('%s=%s' % (variant, value1))
+        variant2 = spack.spec.Spec('%s=%s' % (variant, value2))
+        msg = "'%s' required multiple values for single-valued variant %s" % (pkg,
+                                                                              variant)
+        msg += "\n  requested %s and %s" % (variant1, variant2)
+        raise UnsatisfiableSpecError(msg)
```

Comment:
I think all these error rules can be reduced to one method in `SpecBuilder`, where if you see anything prefixed with `error_`, you call `error(self, *args)`.  This would allow you to move the format strings to the logic program, which will make it much easier to keep them in sync with their args.  I worry that args will change and the format strings won't be updated (as @alalazo already saw in the review).

So, for the first one here you could replace this in `concretize.lp`:

```prolog
error_version_unsatisfiable(Package, Constraint) :-
    ...
```

with this:

```prolog
error_version_unsatisfiable(Package, Constraint, "No valid version for '{}' satisfies '@{}'") :-
    ...
```

And use `args[2].format(args[:2])` on the python side to substitute the first two args into the format string (obviously it'd need to account for arbitrary arg counts).

It seems the only special need here is to be able to construct `Spec` objects for some of the messages, so you could introduce some easy-to-parse embedded syntax for that, e.g., say that `Spec(string)` in the prolog means `Spec(<string>)` on the python side.  So:

```prolog
error_multiple_values_sv_variant(
    Package, Variant, Value1, Value2, 
    "'{0}' required multiple values for single-valued variant {1}\n   requested Spec({1}={2}) Spec({1}={3})"
)
```
I think it should be easy to parse out `Spec()` from the messages and construct the `Spec` objects on the python side.

You could do all this with just one `error()` predicate, though we still need a way to prioritize them in the minimization. I have a suggestion for that below.